renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -246,6 +271,248 @@ void D3D12DebugManager::PixelHistoryCopyPixel(ID3D12GraphicsCommandListX *cmd,
+  {
+    m_pDevice->GetQueue()->GetCommandData()->m_ActionCallback = this;
+  }
+
+  virtual ~D3D12PixelHistoryCallback()
+  {
+    m_pDevice->GetQueue()->GetCommandData()->m_ActionCallback = NULL;
+  }
+
+protected:
+  // Update the given scissor to just the pixel for which pixel history was requested.
+  void ScissorToPixel(const D3D12_VIEWPORT &view, D3D12_RECT &scissor)
+  {
+    float fx = (float)m_CallbackInfo.x;
+    float fy = (float)m_CallbackInfo.y;
+    float y_start = view.TopLeftY;
+    float y_end = view.TopLeftY + view.Height;
+    if(view.Height < 0)
+    {
+      y_start = view.TopLeftY + view.Height;
+      y_end = view.TopLeftY;
+    }
+
+    if(fx < view.TopLeftX || fy < y_start || fx >= view.TopLeftX + view.Width || fy >= y_end)
+    {
+      scissor.left = scissor.top = scissor.right = scissor.bottom = 0;
+    }
+    else
+    {
+      scissor.left = m_CallbackInfo.x;
+      scissor.top = m_CallbackInfo.y;
+      scissor.right = scissor.left + 1;
+      scissor.bottom = scissor.top + 1;
+    }
+  }
+
+  // Intersects the originalScissor and newScissor and writes intersection to the newScissor.
+  // newScissor always covers a single pixel, so if originalScissor does not touch that pixel
+  // returns an empty scissor.
+  void IntersectScissors(const D3D12_RECT &originalScissor, D3D12_RECT &newScissor)
+  {
+    RDCASSERT(newScissor.right == newScissor.left + 1);
+    RDCASSERT(newScissor.bottom == newScissor.top + 1);
+    if(originalScissor.left > newScissor.left || originalScissor.right < newScissor.right ||
+       originalScissor.top > newScissor.top || originalScissor.bottom < newScissor.bottom)
+    {
+      // Scissor does not touch our target pixel, make it empty
+      newScissor.left = newScissor.top = newScissor.right = newScissor.bottom = 0;
+    }
+  }
+
+  // ModifyPSOForStencilIncrement modifies the provided pipeDesc, by disabling depth test
+  // and write, stencil is set to always pass and increment, scissor is set to scissor around
+  // the target pixel, and all color modifications are disabled.
+  // Optionally disables other tests like culling, depth bounds.
+  void ModifyPSOForStencilIncrement(uint32_t eid, D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC &pipeDesc,
+                                    bool disableTests)
+  {
+    pipeDesc.DepthStencilState.DepthEnable = FALSE;
+    pipeDesc.DepthStencilState.DepthWriteMask = D3D12_DEPTH_WRITE_MASK_ZERO;
+
+    if(disableTests)
+    {
+      pipeDesc.RasterizerState.CullMode = D3D12_CULL_MODE_NONE;
+      pipeDesc.RasterizerState.DepthClipEnable = FALSE;
+      pipeDesc.DepthStencilState.DepthBoundsTestEnable = FALSE;
+    }
+
+    // TODO: Get from callbackinfo/pixelhistoryresources?
+    pipeDesc.DSVFormat = DXGI_FORMAT_D32_FLOAT_S8X24_UINT;
+
+    // TODO: If the original depth buffer doesn't have stencil, this will not work as expected.
+    // Does this need to create a new depth/stencil to work correctly in that case?
```

Comment:
Yes this will be one of the requirements of the pixel history, to patch in a stencil buffer as needed even to the extent of converting to a whole new depth buffer format.

Reply:
Updated the TODO comment to indicate that it's definitely needed.